src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -126,20 +129,30 @@ class RsInferenceContext(
         }
     }
 
-    fun infer(fn: RsFunction): RsInferenceResult {
-        extractParameterBindings(fn)
+    fun infer(element: RsInferenceContextOwner): RsInferenceResult {
+        if (element is RsFunction) {
+            extractParameterBindings(element)
+            val fctx = RsFnInferenceContext(this, element.returnType)
+            element.block?.let { fctx.inferFnBody(it) }
+        } else {
+            val (retTy, expr) = when (element) {
+                is RsConstant -> element.typeReference?.type to element.expr
+                is RsArrayType -> TyInteger(TyInteger.Kind.usize) to element.expr
+                is RsVariantDiscriminant -> TyInteger(TyInteger.Kind.isize) to element.expr
+                else -> error("Type inference is not implemented for PSI element of type " +
+                        "`${element.javaClass}` that implement `RsInferenceContextOwner`")
+            }
+            if (expr != null) {
+                RsFnInferenceContext(this, retTy ?: TyUnknown).inferLambdaBody(expr)
```

Comment:
Should we rename `RsFnInferenceContex` to ` RsInferenceContex`?

And we probably can `drop` Rs prefix here, because this is IJ-Rust specific thing and unlikely to clash with something else.

Reply:
The name inspired by rustc (`FnCtxt`)

Reply:
The name inspired by rustc (FnCtxt). We have 2 types: `RsInferenceContext` and `RsFnInferenceContext`, where `RsFnInferenceContext` creates for something that have a return type. For example, for each lambda expr. `RsInferenceContext` is single for each function/constant. So I can't think of a better name for it. The prefix `Rs` exists because there is `InferenceContext` class in JDT and javac.